### PR TITLE
discoveryfeed: Pick one of the clickbaity titles to return

### PR DESCRIPTION
**Pull request by smspillaz**
https://phabricator.endlessm.com/T16620

### CODE REVIEW

File: search-provider/eks-discovery-feed-provider.c [hunk 1]
```diff
@@ -230,6 +261,20 @@ article_card_descriptions_cb (GObject *source,
       add_key_value_pair_from_model_to_variant (model, &builder, "thumbnail-uri");
       add_key_value_pair_from_model_to_variant (model, &builder, "ekn-id");
 
+      /* Select a random clickbait title and add that */
+      g_autoptr(GVariant) clickbaityTitleVariant;
+      g_object_get (model, "clickbaity-titles", &clickbaityTitleVariant, NULL);
+      g_autoptr(GPtrArray) clickbaityTitlesArray = variant_array_string_to_ptr_array(clickbaityTitleVariant,
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Missing space before paren

File: search-provider/eks-discovery-feed-provider.c [hunk 2]
```diff
@@ -230,6 +261,20 @@ article_card_descriptions_cb (GObject *source,
       add_key_value_pair_from_model_to_variant (model, &builder, "thumbnail-uri");
       add_key_value_pair_from_model_to_variant (model, &builder, "ekn-id");
 
+      /* Select a random clickbait title and add that */
+      g_autoptr(GVariant) clickbaityTitleVariant;
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Don't use camel case for C code

File: search-provider/eks-discovery-feed-provider.c [hunk 3]
```diff
@@ -230,6 +261,20 @@ article_card_descriptions_cb (GObject *source,
...
+      g_object_get (model, "clickbaity-titles", &clickbaityTitleVariant, NULL);
+      g_autoptr(GPtrArray) clickbaityTitlesArray = variant_array_string_to_ptr_array(clickbaityTitleVariant,
+                                                                                     ESTIMATED_CLICKBAITY_TITLES);
+
+      if (clickbaityTitlesArray->len)
+        {
+          add_key_value_pair_to_variant (&builder,
+                                         "clickbaity_title",
+                                         g_ptr_array_index(clickbaityTitlesArray,
+                                                           rand () % clickbaityTitlesArray->len));
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: It feels wrong to go from variant to pointer array to variant again like this.
You can just use `g_variant_get_child()`, or `g_variant_get_child_value()` on the original variant instead.

File: search-provider/eks-discovery-feed-provider.c [hunk 4]
```diff
@@ -278,9 +377,15 @@ handle_article_card_descriptions (EksDiscoveryFeedDatabaseContentProvider *skele
     g_variant_builder_add (&tags_match_any_builder, "s", "EknArticleObject");
     GVariant *tags_match_any = g_variant_builder_end (&tags_match_any_builder);
 
+    GVariantBuilder tags_match_all_builder;
+    g_variant_builder_init (&tags_match_all_builder, G_VARIANT_TYPE ("as"));
+    g_variant_builder_add (&tags_match_all_builder, "s", "EknHasDiscoveryFeedTitle");
+    GVariant *tags_match_all = g_variant_builder_end (&tags_match_all_builder);
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Is this not leaked?

File: search-provider/eks-discovery-feed-provider.c [hunk 5]
```diff
@@ -181,19 +183,80 @@ underscorify (const gchar *string)
...
+static const gchar *
+select_random_string_from_ptr_array(GPtrArray *array)
+{
+  if (array->len)
+    return g_ptr_array_index(array, rand () % array->len);
+
+  return NULL;
+}
+
+static const unsigned int DISCOVERY_FEED_SET_CUSTOM_TITLE = 1 << 0;
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Make this a typedeffed enum and add a `NONE` value that maps to `0`

File: search-provider/eks-discovery-feed-provider.c [hunk 6]
```diff
@@ -244,10 +307,46 @@ article_card_descriptions_cb (GObject *source,
       g_variant_builder_open (&builder, G_VARIANT_TYPE ("a{ss}"));
 
       EkncContentObjectModel *model = l->data;
+      guint flags = 0;
+
+      /* Examine the discovery-feed-content object first and set flags
+       * for things that we've overridden */
+      g_autoptr(GVariant) discoveryFeedContentVariant;
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Don't use camel case names in C

File: search-provider/eks-discovery-feed-provider.c [hunk 7]
```diff
@@ -244,10 +307,46 @@ article_card_descriptions_cb (GObject *source,
...
+      GVariantIter discoveryFeedContentIter;
+      g_variant_iter_init (&discoveryFeedContentIter,
+                           discoveryFeedContentVariant);
+
+      gchar *key;
+      GVariant *value;
+
+      while (g_variant_iter_loop (&discoveryFeedContentIter, "{sv}", &key, &value))
+        {
+          if (g_strcmp0(key, "blurbs") == 0)
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Missing space before paren (multiple occurrences)

File: search-provider/eks-discovery-feed-provider.c [hunk 8]
```diff
@@ -244,10 +307,46 @@ article_card_descriptions_cb (GObject *source,
...
+            {
+              g_autoptr(GPtrArray) array = variant_array_variant_to_ptr_array(value,
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: I think you can avoid putting this into a `GPtrArray` altogether, and simply select randomly a child variant from the array using `g_variant_get_child_value ()` or a similar function

File: search-provider/eks-discovery-feed-provider.c [hunk 9]
```diff
@@ -181,19 +183,80 @@ underscorify (const gchar *string)
...
+{
+  GVariantIter iter;
+  GVariant *value;
+
+  GPtrArray *array = g_ptr_array_new_full (reserved_size, g_free);
+
+  g_variant_iter_init (&iter, variant);
+  while (g_variant_iter_loop (&iter, "v", &value))
+    {
+      g_ptr_array_add (array, g_strdup(g_variant_get_string(value, NULL)));
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: You could use `g_variant_dup_string()`, but as I note below, I think this function should go away.

File: search-provider/eks-discovery-feed-provider.c [hunk 10]
```diff
@@ -7,6 +7,8 @@
 
 #include <eos-knowledge-content.h>
 #include <eos-shard/eos-shard-shard-file.h>
+
+#include <stdlib.h>
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: If this is for `rand()`, I think you can just use `g_random_int_range()` instead

File: search-provider/eks-discovery-feed-provider.c (hunk 4 above)
Comment (smspillaz) [inline search-provider/eks-discovery-feed-provider.c]: It seems like the constructor for `EkncQueryObject` consumes the value. I put it behind g_autoptr and got warnings about the ref count going below 1.

File: search-provider/eks-discovery-feed-provider.c (hunk 8 above)
Comment (smspillaz) [inline search-provider/eks-discovery-feed-provider.c]: Yeah this is a good point - thanks for the tip on the API.

File: search-provider/eks-discovery-feed-provider.c (hunk 6 above)
Comment (smspillaz) [inline search-provider/eks-discovery-feed-provider.c]: -ETOOMANYLANGUAGES

File: search-provider/eks-discovery-feed-provider.c (hunk 5 above)
Comment (smspillaz) [inline search-provider/eks-discovery-feed-provider.c]: Done

File: search-provider/eks-discovery-feed-provider.c (hunk 10 above)
Comment (smspillaz) [inline search-provider/eks-discovery-feed-provider.c]: Good point, done.

File: search-provider/eks-discovery-feed-provider.c [hunk 11]
```diff
@@ -181,19 +182,44 @@ underscorify (const gchar *string)
...
-  g_variant_builder_add (builder, "s", underscore_key);
-  g_variant_builder_add (builder, "s", value);
-  g_variant_builder_close (builder);
+  add_key_value_pair_to_variant (builder, underscore_key, value);
+}
+
+static gchar *
+select_random_string_from_variant(GVariant *variant)
+{
+  gsize size = g_variant_n_children(variant);
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Missing space before paren

File: search-provider/eks-discovery-feed-provider.c [hunk 12]
```diff
@@ -181,19 +182,44 @@ underscorify (const gchar *string)
...
+  gint index = g_random_int_range (0, size);
+  /* Note that the extracted child_variant here is of type 'v', we still
+   * need to unwrap it one more time to get the string */
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Optional: you could probably use `g_variant_get_child()` to make this even more compact

File: search-provider/eks-discovery-feed-provider.c [hunk 13]
```diff
@@ -181,19 +182,44 @@ underscorify (const gchar *string)
...
+  g_autoptr(GVariant) child_variant = g_variant_get_child_value(variant, index);
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Missing space before paren (on every line until the end of the function)

File: search-provider/eks-discovery-feed-provider.c [hunk 14]
```diff
@@ -181,19 +182,44 @@ underscorify (const gchar *string)
...
   g_autofree gchar *underscore_key = underscorify (key);
   g_object_get (model, key, &value, NULL);
-  g_variant_builder_add (builder, "s", underscore_key);
-  g_variant_builder_add (builder, "s", value);
-  g_variant_builder_close (builder);
+  add_key_value_pair_to_variant (builder, underscore_key, value);
+}
+
+static gchar *
+select_random_string_from_variant(GVariant *variant)
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Missing space after function name

File: search-provider/eks-discovery-feed-provider.c [hunk 15]
```diff
@@ -244,10 +270,44 @@ article_card_descriptions_cb (GObject *source,
...
+      GVariantIter discovery_feed_content_iter;
+      g_variant_iter_init (&discovery_feed_content_iter,
+                           discovery_feed_content_variant);
+
+      gchar *key;
+      GVariant *value;
+
+      while (g_variant_iter_loop (&discovery_feed_content_iter, "{sv}", &key, &value))
+        {
+          if (g_strcmp0(key, "blurbs") == 0)
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: Missing space before paren

File: search-provider/eks-discovery-feed-provider.c (hunk 12 above)
Comment (smspillaz) [inline search-provider/eks-discovery-feed-provider.c]: I wasn't able to get this to work :( I tried using something like `g_variant_get_child(variant, index, "(s)", &str, NULL)`, but GVariant was complaining that the underlying value is of type `v` and wouldn't unbox the inner string.

**Comment by smspillaz** [comment]
Other than that, style comments should be fixed.

File: search-provider/eks-discovery-feed-provider.c [hunk 16]
```diff
@@ -181,19 +182,44 @@ underscorify (const gchar *string)
...
+  add_key_value_pair_to_variant (builder, underscore_key, value);
+}
+
+static gchar *
+select_random_string_from_variant (GVariant *variant)
+{
+  gsize size = g_variant_n_children (variant);
+  gint index = g_random_int_range (0, size);
+  /* We need to unwrap the variant and then the inner string first */
+  g_autoptr(GVariant) child_variant = g_variant_get_child_value(variant, index);
```
Comment (cosimoc) [inline search-provider/eks-discovery-feed-provider.c]: This line and the following one still have a missing space before paren.

Comment (smspillaz) [inline search-provider/eks-discovery-feed-provider.c]: 😭

Comment (smspillaz) [inline search-provider/eks-discovery-feed-provider.c]: Sorry about the noise :/ Should be fixed now

**Pull request merged**